pkg/virtctl/usbredir/usbredir.go

```
@@ -0,0 +1,217 @@
+
+	// Configure in/out and start stream with websocket
+	k8ResChan := make(chan error)
+	go func() {
+		streamOpts := kubecli.StreamOptions{
+			In:  pipeInReader,
+			Out: pipeOutWriter,
+		}
+		k8ResChan <- usbredirVMI.Stream(streamOpts)
+	}()
+
+	port, err := getFreePort()
+	if err != nil {
+		return fmt.Errorf("Could not get local TCP port %s: %s", vmiArg, err.Error())
+	}
+
+	// Connect to usbredirserver and forward data to/from websocket
+	usbredirDoneChan := make(chan struct{}, 1)
+	streamResChan := make(chan error)
+	streamStop := make(chan error)
+	go func() {
+		start := time.Now()
+
+		// Can take a few ms to usbredirserver to run, so we try for a few seconds
+		var usbredirConn net.Conn
+		count := 5
+		for {
+			usbredirConn, err = net.Dial("tcp", fmt.Sprintf(":%v", port))
+			if err == nil {
+				break
+			}
+
+			if count > 0 && strings.Contains(err.Error(), "connection refused") {
+				time.Sleep(time.Second)
+				count -= 1
+			} else {
+				glog.Infof("Failed to connect to usbredirserver: %v", err)
+				streamResChan <- err
+				return
+			}
+		}
+		defer usbredirConn.Close()
+
+		glog.V(2).Infof("Connected to usbredirserver %v", time.Now().Sub(start))
+
+		// write to usbredirserver from pipeOutReader
+		go func() {
+			_, err := io.Copy(usbredirConn, pipeOutReader)
+			streamStop <- err
+		}()
+
+		// read from usbredirserver towards pipeInWriter
+		go func() {
+			_, err := io.Copy(pipeInWriter, usbredirConn)
+			streamStop <- err
+		}()
+
+		// Wait for usbredirserver to complete
+		<-usbredirDoneChan
+		streamResChan <- err
+	}()
+
+	// execute usbredirserver
+	usbredirResultChan := make(chan error)
+	go func() {
+		defer close(usbredirDoneChan)
+
+		args := []string{}
+		port_arg := fmt.Sprintf("-p %v", port)
+		args = append(args, port_arg, usbdeviceArg)
+
+		bin := USBREDIRSERVER
+		if _, err := exec.LookPath(bin); err != nil {
+			usbredirResultChan <- fmt.Errorf("could not find %s binary in $PATH", USBREDIRSERVER)
```

Maybe we can move this check to the beginning of the process and fail faster?

---

Done! Thanks